test_conformance/api/test_kernel_attributes.cpp

```diff
@@ -0,0 +1,343 @@
+AttributePermutations work_vect_tests;
+AttributePermutations reqd_work_tests;
+
+AttributePermutations vect_work_reqd_tests;
+AttributePermutations work_reqd_vect_tests;
+AttributePermutations reqd_vect_work_tests;
+
+
+/* Generate a vector with vec_type_hint(<data_type>) so that it can be used to
+ * generate different kernels */
+static KernelAttributes generate_vec_type_hint_data(cl_device_id deviceID)
+{
+    KernelAttributes vec_type_hint_data;
+    std::vector<std::string> vector_types = { "char",   "uchar", "short",
+                                              "ushort", "int",   "uint",
+                                              "float" };
+    if (gHasLong)
+    {
+        vector_types.push_back("long");
+        vector_types.push_back("ulong");
+    }
+    if (device_supports_half(deviceID))
+    {
+        vector_types.push_back("half");
+    }
+    if (device_supports_double(deviceID))
+    {
+        vector_types.push_back("double");
+    }
+
+    const auto vector_sizes = { "2", "3", "4", "8", "16" };
+    for (auto type : vector_types)
+    {
+        for (auto size : vector_sizes)
+        {
+            vec_type_hint_data.push_back("vec_type_hint(" + type + size + ")");
+        }
+    }
+    return vec_type_hint_data;
+}
+
+
+struct WorkGroupDimensions
+{
+    int x;
+    int y;
+    int z;
+};
+
+/* Generate vectors to store reqd_work_group_size(<dimensions>) and
+ * work_group_size_hint(<dimensions>) so that they can be used to generate
+ * different kernels */
+static KernelAttributes generate_reqd_work_group_size_data(
+    const std::vector<WorkGroupDimensions>& work_group_dimensions)
+{
+    KernelAttributes reqd_work_group_size_data;
+    for (auto dimension : work_group_dimensions)
+    {
+        reqd_work_group_size_data.push_back(
+            "reqd_work_group_size(" + std::to_string(dimension.x) + ","
+            + std::to_string(dimension.y) + "," + std::to_string(dimension.z)
+            + ")");
+    }
+    return reqd_work_group_size_data;
+}
+
+static KernelAttributes generate_work_group_size_data(
+    const std::vector<WorkGroupDimensions>& work_group_dimensions)
+{
+    KernelAttributes work_group_size_hint_data;
+    for (auto dimension : work_group_dimensions)
+    {
+        work_group_size_hint_data.push_back(
+            "work_group_size_hint(" + std::to_string(dimension.x) + ","
+            + std::to_string(dimension.y) + "," + std::to_string(dimension.z)
+            + ")");
+    }
+    return work_group_size_hint_data;
+}
+
+/* Populate the Global Vectors which store individual Kernel Attributes */
+static void populate_single_attribute_tests(
+    /* Vectors to store the different data that fill the attributes */
+    const KernelAttributes& vec_type_hint_data,
+    const KernelAttributes& work_group_size_hint_data,
+    const KernelAttributes& reqd_work_group_size_data)
+{
+    for (auto vector_test : vec_type_hint_data)
+    {
+        /* Initialise vec_type_hint attribute tests */
```

Comment:
Very nitty style point, but I think we should try to consistently use C++ comments in most places (`//`), as we're generally moving the codebase to be more canonically C++.